File: internal/gamelogic/platformer/level.go
package platformer

import (
	"github.com/1001bit/OnlineCanvasGames/internal/gamelogic"
	"github.com/1001bit/OnlineCanvasGames/internal/mathobjects"
	"github.com/1001bit/OnlineCanvasGames/pkg/concurrent"
)

type LevelConfig struct {
	PlayerSpeed    float64 `json:"playerSpeed"`
	PlayerJump     float64 `json:"playerJump"`
	PlayerGravity  float64 `json:"playerGravity"`
	PlayerFriction float64 `json:"playerFriction"`
}

type Level struct {
	// [rectID]rect
	players map[int]*Player
	blocks  map[int]*Block

	// userID[data]
	playersData concurrent.ConcurrentMap[int, *PlayerData]

	config LevelConfig

	serverTPS float64
	clientTPS float64
}

func NewPlatformerLevel() *Level {
	var (
		config = LevelConfig{
			PlayerSpeed:    3,
			PlayerJump:     5,
			PlayerGravity:  0.03,
			PlayerFriction: 0.3,
		}

		serverTPS = 20.0
		clientTPS = 50.0
	)

	level := &Level{
		players: make(map[int]*Player),
		blocks:  make(map[int]*Block),

		playersData: concurrent.MakeMap[int, *PlayerData](),

		config: config,

		serverTPS: serverTPS,
		clientTPS: clientTPS,
	}

	block := NewBlock(0, 500, 1000, 100)
	level.blocks[10] = block

	return level
}

func (l *Level) Tick(dtMs float64, writer gamelogic.RoomWriter) {
	movedPlayers := make(map[int]mathobjects.Vector2[float64])

	// Controls
	playersData, rUnlockFunc := l.playersData.GetMapForRead()
	for _, playerData := range playersData {
		playerData.ControlPlayer(l.config.PlayerSpeed, l.config.PlayerJump)
	}
	rUnlockFunc()

	// Physics
	for rectID, player := range l.players {
		startPos := player.GetPosition()

		// Forces
		player.ApplyGravity(l.config.PlayerGravity, dtMs)
		player.ApplyFriction(l.config.PlayerFriction)

		// Collisions and displacement
		player.SetCollisionDir(mathobjects.None)

		// Horizontal
		for _, block := range l.blocks {
			dir := player.DetectHorizontalCollision(block, dtMs)
			if dir != mathobjects.None {
				player.ResolveCollision(block, dir)
				break
			}
		}
		player.Position.X += player.velocity.X * dtMs

		// Vertical
		for _, block := range l.blocks {
			dir := player.DetectVerticalCollision(block, dtMs)
			if dir != mathobjects.None {
				player.ResolveCollision(block, dir)
				break
			}
		}
		player.Position.Y += player.velocity.Y * dtMs

		if startPos != player.GetPosition() {
			movedPlayers[rectID] = player.GetPosition()
		}
	}

	// Level Update Message
	writer.GlobalWriteMessage(NewLevelUpdateMessage(movedPlayers))
}

File: web/typescript/gameassets/2/level.ts
interface LevelConfig {
    playerSpeed: number;
    playerJump: number;
    playerGravity: number;
    playerFriction: number;
}

class Level {
    private blocks: Map<number, Block>
    private interpolatedPlayers: Map<number, InterpolatedPlayer>
    private kinematicPlayers: Map<number, KinematicPlayer>

    private config: LevelConfig;
    private playerRectID: number

    private fixedTicker: FixedTicker;
    private serverTPS: number;
    private serverAccumulator: number;

    constructor(){
        this.blocks = new Map()
        this.interpolatedPlayers = new Map()
        this.kinematicPlayers = new Map()

        this.config = {
            playerSpeed: 0,
            playerJump: 0,
            playerGravity: 0,
            playerFriction: 0,
        },
        this.playerRectID = 0

        this.fixedTicker = new FixedTicker(10)
        
        this.serverTPS = 0
        this.serverAccumulator = 0
    }

    setConfig(config: LevelConfig){
        this.config = config
    }

    setPlayerRectID(id: number){
        this.playerRectID = id
    }

    setTPS(client: number, server: number){
        this.serverTPS = server
        this.fixedTicker.setTPS(client)
    }

    createPlayerRectangle(serverRect: AbstractPlayer, rectID: number){
        if(this.interpolatedPlayers.has(rectID) || this.kinematicPlayers.has(rectID)){
            return
        }

        let rectangle: RectangleShape | undefined;

        if(rectID == this.playerRectID){
            const rect = new KinematicPlayer(serverRect)

            this.kinematicPlayers.set(rectID, rect)
            rectangle = new RectangleShape(rect)
        } else {
            const rect = new InterpolatedPlayer(serverRect)

            this.interpolatedPlayers.set(rectID, rect)
            rectangle = new RectangleShape(rect)
        }

        return rectangle
    }

    disconnectPlayer(rectId: number){
        this.interpolatedPlayers.delete(rectId)
        this.kinematicPlayers.delete(rectId)
    }

    createBlockRectangle(serverRect: AbstractBlock, rectID: number){
        if(this.blocks.has(rectID)){
            return
        }

        const rect = new Block(serverRect)
        this.blocks.set(rectID, rect)

        const rectangle = new RectangleShape(rect)
        return rectangle
    }

    tick(dt: number, controls: Controls){
        this.serverAccumulator += dt

        // interpolate interpolated players
        const interpolatedAlpha = Math.min(this.serverAccumulator/(1000/this.serverTPS), 1)
        for (const [_, player] of this.interpolatedPlayers){
            player.interpolate(interpolatedAlpha)
        }

        // update kinematic players
        this.fixedTicker.update(dt, fixedDT => {
            for(const [rectID, player] of this.kinematicPlayers){
                // update interpolation
                player.updateStartPos()

                // Control
                if(rectID == this.playerRectID){
                    player.control(this.config.playerSpeed, this.config.playerJump, controls)
                }

                // Forces
                player.applyGravity(this.config.playerGravity, fixedDT)
                player.applyFriction(this.config.playerFriction)

                // Collisions and movement
                player.setCollisionDir(Direction.None)

                // Horizontal
                for(const [_, block] of this.blocks){
                    const dir = player.detectHorizontalCollision(block, fixedDT)
                    if(dir != Direction.None){
                        player.resolveCollision(block, dir)
                        break
                    }
                }
                player.applyHorizontalVelToPos(fixedDT)

                // Vertical
                for(const [_, block] of this.blocks){
                    const dir = player.detectVerticalCollision(block, fixedDT)
                    if(dir != Direction.None){
                        player.resolveCollision(block, dir)
                        break
                    }
                }
                player.applyVerticalVelToPos(fixedDT)
            }
        })

        // interpolate kinematic players
        const kinematicAlpha = this.fixedTicker.getAlpha()
        for (const [_, player] of this.kinematicPlayers){
            player.interpolate(kinematicAlpha)
        }
    }

    handlePlayerMovement(moved: {}, correct: boolean){
        // update interpolated rects interpolation
        this.serverAccumulator = 0
        
        for (const [_, player] of this.interpolatedPlayers){
            player.updateStartPos()
        }
        
        // set position
        for (const [key, val] of Object.entries(moved)){
            const rectID = Number(key)
            const pos = val as {x: number, y: number}

            const interpolated = this.interpolatedPlayers.get(rectID)
            if(interpolated){
                interpolated.setTargetPos(pos.x, pos.y)
                continue
            }

            const kinematic = this.kinematicPlayers.get(rectID)
            if(kinematic){
                if(correct){
                    kinematic.correctDivergence(pos.x, pos.y)
                }
            }
        } 
    }
}